File: python_agent/submission/agent.py
from __future__ import annotations

import random
from typing import Tuple

from ..controller import Controller
from ..move import Move
from ..custom_logger import log


def make_move(controller: Controller) -> Tuple[bool, Move]:
    """Pick a random valid move, mirroring the default C++ submission."""
    log('Hi from make_move function')
    time_ms = controller.get_time_ms()
    log(f'Time remaining: {time_ms} ms')

    board = controller.get_current_board()
    valid_moves = board.get_valid_moves()
    if not valid_moves:
        dummy_move = Move(0, 0, True)
        return False, dummy_move

    move = random.choice(valid_moves)
    requires_more = controller.make_move(move)
    log(f'Made move: {move}, requires more: {requires_more}')
    return requires_more, move


__all__ = ["make_move"]

File: cpp_agent/src/submission/agent.cpp
#include "agent.h"
#include "customLogger.h"
#include <cstdlib>
#include <ctime>
#include <iostream>
#include <utility>

void Agent::init() { srand(time(NULL)); }

// Picks one random move, applies it via Controller, returns {requiresContinuation, move}
std::pair<bool, Move> makeMove(Controller &controller) {
    Log("Hi from makeMove function");
    Log(controller.getTimeMs());
    
    Board board = controller.getCurrentBoard();
    auto validMoves = board.getValidMoves();
    if (validMoves.empty()) {
        return {false, Move{0,0,true}}; // no moves to play
    }
    const Move &m = validMoves[rand() % validMoves.size()];
    bool cont = controller.makeMove(m);
    // log my move
    string logMessage = "Agent made move: " + to_string(m.row) + " " + to_string(m.col) + " " + to_string(m.isHorizontal);
    Log(logMessage);
    return {cont, m};
}

void Agent::run() {
    //play until the board is full or no moves remain.
    while (!controller->getCurrentBoard().isCompleted()) {
        while (true) {
            auto [requiresMoreMoves, move] = makeMove(*controller);
            if (!requiresMoreMoves || controller->getCurrentBoard().isCompleted()) break;
        }
        if (controller->getCurrentBoard().isCompleted()) {
            break;
        }
    }
}
